[RDY][CR] Rework inventory columns

codemime opened the pull request:
They definitely deserve a class of their own. It gives additional flexibility for managing them and makes the code cleaner. In the future this will enable us to have a shared functionality with the pick-up menu and support filtering by keyboard input (along with some other features).

TODO:
- [x] The column that shows selected items should become a descendant of `inventory_column`, not an ugly duplicate
- [x] Remove `first_item` and `second_item`
- [x] Adaptive width of columns
- [x] Ground and vehicle items in a separate column (only when there's free room on the screen)
- [x] Ability to select many items at once (I suppose via category mode).

EDIT:
### Summary:

Almost all the code of inventory columns was rewritten. Now it's much cleaner and more object-oriented. Due to length I'll omit the list of code changes, they can be seen in the code itself, but I'll describe observable changes in detail.

The most significant feature is the actual multidrop. You can drop a whole category within a couple of key-presses (i.e. get undressed). Number of items to drop (or NONE) is shown. Also, markers are colorized and no more red color for categories:
![undress](https://cloud.githubusercontent.com/assets/11905613/16157180/863c9c84-34c8-11e6-9bc4-890b798f91ee.gif)

You're free to use category selection in any mode, but those that require only one item, won't allow you to select more.

If you drop too many items at once, the last column (ITEMS TO DROP) becomes selectable and you can inspect it (see the old behavior below): 
![long_drop](https://cloud.githubusercontent.com/assets/11905613/16157279/ecd471ba-34c8-11e6-8ac6-302e2d9c118d.gif)
Here's the old behavior:
![old_list_overflow](https://cloud.githubusercontent.com/assets/11905613/16157289/f87cff00-34c8-11e6-9830-111805a268d1.png)

Selected column is always the topmost so that nothing's trimmed and can be fully read. Inactive columns don't show their selection (it was misleading):
![select](https://cloud.githubusercontent.com/assets/11905613/16157335/2393a2d4-34c9-11e6-869c-c60e23ba36ba.gif)
The rightmost column is prepared for the longest item, so there's no "jerking":
![space](https://cloud.githubusercontent.com/assets/11905613/16157652/7b344ae2-34ca-11e6-818f-63ee5c1b050c.gif)

Here's the old appearance:
![old_trim](https://cloud.githubusercontent.com/assets/11905613/16157384/6682f57c-34c9-11e6-9a88-1f5960b67a62.png)
![no_drop_list](https://cloud.githubusercontent.com/assets/11905613/16157594/38cf884c-34ca-11e6-81bf-7e4cd1094390.png)

Skimming through categories works mostly as usual, but was also improved. It will not skip pages in case of long categories (note TOOLS) and Page x/y is shown under each column, but only if it's necessary (more than one page):
![categories](https://cloud.githubusercontent.com/assets/11905613/16157364/46860fe8-34c9-11e6-9bda-af6645545bd8.gif)

Page up/page down are now available:
![page_updown](https://cloud.githubusercontent.com/assets/11905613/16157420/853621f6-34c9-11e6-8094-ae5f91c5aba3.png)

There's a short static summary:
![middle_column](https://cloud.githubusercontent.com/assets/11905613/16157437/9b4ed960-34c9-11e6-8c82-eeeb41e151b9.png)

Selection mode is now correctly "remembered" when you select an item and then go back to the inventory (actually it doesn't change, because the object stays the same and doesn't get recreated).
Also, there are some more small changes that I probably forgot to mention.

I plan more extensive clean up for `inventory_ui.cpp` in further PRs.

EDIT: Fixes #16504.

codemime commented:
The PR is ready for review / testing.

Coolthulhu commented on src/inventory_ui.cpp:
```diff
@@ -781,19 +967,28 @@ int inventory_selector::execute_pick( const std::string &title, const int positi
+                return selection.front()->item_pos;
+            } else {
+                const std::string msg( _( "Please pick only one item." ) );
+                popup( msg , PF_GET_KEY );
```
This shouldn't really happen.
When the selector is in single-select mode, it should not look like it is in multi-select mode and should not allow selecting multiple items. Making it default to the first one is fine, as long as it displays only the first one as selected.

Also, "please" in user commands is a bit too polite. It sounds like we're giving the user a choice, which isn't the case. "Only one item can be selected" or something like that.

codemime replied:
Makes sense.

Coolthulhu commented on src/inventory_ui.cpp:
```diff
@@ -879,66 +1110,73 @@ void inventory_selector::execute_compare( const std::string &title )
+        }
+        refresh_all();
+        res = inventory_item_menu( pos );
+    } while( std::find( stoppers.begin(), stoppers.end(), res ) == stoppers.end() );
```
Would be cleaner if you just made the `stoppers` a `set` and used `count` here.

codemime replied:
Totally agreed.

Coolthulhu commented on src/inventory_ui.cpp:
```diff
@@ -767,10 +876,87 @@ void inventory_selector::remove_dropping_items( player &u ) const
+{
+    assert( new_column != nullptr );
+
+    std::unique_ptr<inventory_column> column( new_column );
```
This looks risky. Why are you wrapping a non-unique ptr in a unique ptr? Why not pass unique one as argument instead?
Passing unique ptr would make the intention clear and prevent possible bugs resulting from someone trying to keep the column despite inserting it.

Also, all usages of the function are in the form:
```
insert_column( iter, uniqueptr.release() );
```

You could just change the `ptr.release()` into `std::move( ptr )`.

codemime replied:
> Passing unique ptr would make the intention clear

That was my first intention and I implemented it. I had to use casting to achieve polymorphic unique pointers. That didn't look particularly elegant, so I abandoned the idea. But you probably right about the safety and clearness.

codemime commented:
All set. Cleaned up and fixed a couple of bugs in the current `master`: 
![01](https://cloud.githubusercontent.com/assets/11905613/16179639/514c1bde-367c-11e6-9782-d2d329c17250.gif)
Pressing PAGE DOWN. Note that page 4 is never shown because it has lesser items than would be selected. Moreover, PAGE UP and PAGE DOWN behave weird when there's only one page on the screen (i.e. press PAGE DOWN -> the topmost item will be selected and vice versa). The new mechanics is predictable and it feels right.

codemime commented:
Position of inventory columns on the screen is now adaptive. They're aligned to the left of the screen if they occupy less than 2/3 of its width, otherwise they'll occupy all the width evenly. Also, the rounding error for the gaps was taken into account.

codemime commented:
Anything else before it's mergeable? I'm going to PR an improvement that would depend on both this PR and #16766.

Coolthulhu commented on src/inventory_ui.cpp:
```diff
@@ -40,409 +32,697 @@ struct itemstack_or_category {
+        : it( &( slice.first->front() ) ), slice( slice.first ),
+          category( ( cat != nullptr ) ? cat : &it->get_category() ), item_pos( slice.second ) {}
+    inventory_entry( const item *it, int pos, const item_category *cat = nullptr )
+        : it( it ), slice( nullptr ), category( ( cat != nullptr ) ? cat : &it->get_category() ), item_pos( pos ) {}
```
Those defaults are unreadable. If you're going to have ternary operators inside them, it would be a good idea to put each of them on a separate line.

codemime replied:
Agreed. Their formatting is a leftover.

Coolthulhu commented:
* Get a screen size wide enough for 3 columns (mine is 135 columns)
* Get some items - worn and in inventory
* Displays as 2 columns
* Get a lot of items so that one page won't display them all at once
* Displays as 3 columns now, but the middle column is just empty space

codemime commented:
I try but fail to get an unexpected behavior. Can it be this:

> Position of inventory columns on the screen is now adaptive. They're aligned to the left of the screen if they occupy less than 2/3 of its width, otherwise they'll occupy all the width evenly.

Coolthulhu commented:
Pretty sure it's this PR:
Dropping charged items makes them uncharged. Dropping a specific amount of charges doesn't do that.

codemime commented:
I tried to drop a vegetable pie and nobody has eaten it while it was falling. I'll try to rebase this PR against current master and try agian. Perhaps something has happened since then.

Coolthulhu commented:
There is a slight chance some other PR did that, but it only seems to happen when dropping things with the inventory menu - autodrop, advanced inventory, throwing etc. are all unaffected.

codemime commented:
Yes, it's changed after some very recent PR. I'm nailing it down now.

codemime commented:
It's definitely not this PR. The issue is in the `master` branch already.

Coolthulhu commented:
This PR can't be properly tested for as long as this bug exists.

codemime commented:
I found a source of the bug and will PR a fix pretty soon.
